### ISSUE TITLE
Full mode: DPR retrieval baseline runs on the chunks from all dialogue

### ISSUE THREAD

**Issue by luomancs** [salesforce/QAConv#5]
Hi there, 

I tried to reproduce the results of DPR baseline on the full mode dataset. But somehow, from the code, I see that in this baseline, you first generate the embeddings (around 18K chunks) for all the dialogues context, then for each question, you retrieve the chunk for each question from this large corpus. But for BM25 you only retrieve the chunks from the given dialogue or channel. This sounds unfair for two baselines? Please let me know if I missed any code which might clear out my confusion. 

Thank you.

**Comment by jasonwu0731** [comment]
Hi @luomancs,
 
Can you point me to the lines that you are referring to? I remember both methods should call the same function *get_chunks_by_qa*. It will be helpful to print out some numbers if you think it helps. Thanks.

**Comment by luomancs** [comment]
Hi @jasonwu0731, 

According to run_dpr.sh, DPR baselines generates the dense dense representation for the entire corpus (18K+ chunks) first, and then for each question, it retrieve 100 chunks from these entire corpus. After this, in the dpr.py, it calls the get_chunks_by_qa (Line 33) and returns a small set of chunks (chunk_dict) given a dialogue or channel.  But since not every chunk appears in the retrieved 100 chunks, line 36 gives error that chunk_key does not find.

Let me know if I missed something. 

Thanks a lot.

**Comment by jasonwu0731** [comment]
Right, we need to handle the situation when a chunk is not retrieved by DPR and assign a low score like 0 to it.

sth like this 
```
for chunk_i, (chunk_key, chunk_value) in enumerate(chunks_dict.items()):
    if chunk_key in score_dict.keys():
            output_retriever_rank[qa_pair["id"]][chunk_key] = score_dict[chunk_key]
    else:
            output_retriever_rank[qa_pair["id"]][chunk_key] = 0
```
https://github.com/salesforce/QAConv/blob/master/retriever/dpr.py#L35

with this can you get similar numbers as what we have in the paper? If not, can you show us your results and the commands you run? Thanks.

**Comment by jasonwu0731** [comment]
Hi @luomancs 

Thanks for pointing this out. I found that the V.1 version I uploaded last month is not the correct one. I just correct that and now you should have a V1.1 that has a balanced val and tst splits (~3660 each). Please download the dataset folder again you let me know if you still can't replicate the retrieval results reported.

These are the [retrieval files](https://drive.google.com/file/d/1Eyzkan6e9VRU3XkH4FEqu5AX7GgzwPNY/view?usp=sharing) I found (should be the latest one for V1.1), you can take this as reference.

**Comment by luomancs** [comment]
Hi @jasonwu0731 ,

Thanks for updating the new files. By using this, at least I can get the same BM25 results as in the paper, however, I still can not reproduce the DPR results. The new results I had for DPR is  (paper/reproduce):
0.429/0.395 | 0.601/0.563 | 0.661/0.613 | 0.740/0.690

One thing I found in your retrieval files, output_retriever_rank_dpr-wiki_all_tst.json, for some questions, it actually has more than 100 retrieved chunks, however, it does not match with the setting in the code, where the default n_docs is 100 (in dense_retriever.yaml file). 
I guess probably DPR retrieves all chunks (18728) and then in dpr.py, only those chunks appearing in the dialogue/slack are selected for evaluation? 

When I run this command,   
>     python dense_retriever.py model_file=$(pwd)/downloads/checkpoint/retriever/single-adv-hn/nq/bert-base-encoder.cp qa_dataset=qaconv_test ctx_datatsets=[qaconv] encoded_ctx_files=[\"$(pwd)/article_segment_0\"] out_file=../../../qaconv_test_dpr_results.json

it will generates a file called qaconv_test_dpr_results.json, could you also share this file? It might clear out if my guess is right, thanks a lot.

**Comment by shashankneocfc** [comment]
Hi @jasonwu0731 
Following up on the above question, can you please help to share the qaconv_test_dpr_results.json file ?
